### PR TITLE
[WIP] Specialize accumulators

### LINKED ISSUE #63: Specialize FMA Accumulators Properly
If we didn't turn `FixAdd(FixMul(a,b),c)` into `FixFMA(a,b,c)`, then a `Reduce(0)(...){i => i*i}{_+_}` would not have a huge II, since the FixMul would be outside of the reduction loop between RegRead, FixAdd, RegWrite.  But assuming we are to keep this transform, I am working on a specialized accumulator for this case that has as many lanes as II and fuses the lanes in a reduction tree that allows the effective II = 1.

### PR DESCRIPTION
**Pull request by dkoeplin**
@mattfel1 Take a look at this implementation, particularly the accumulator nodes for registers in 
`spatial.node.Accumulator`
Maybe also check if the transformer rules in `spatial.transform.AccumTransformer` look sane to you

If this works for you, we can add backend support for these in scala and chisel

Fixes #63

### CODE REVIEW

File: src/spatial/node/Accumulator.scala [hunk 1]
```diff
@@ -1,5 +1,42 @@
...
+  extends RegAccum[A]
+
+@op case class RegAccumOp[A:Bits](
+    mem: Reg[A],
+    in:  Bits[A],
+    en:  Set[Bit],
+    op:  Accum,
+    first: Bit)
+  extends RegAccum[A]
+
```
Comment (mattfel1) [inline src/spatial/node/Accumulator.scala:51]: Minor housekeeping detail, but does it make sense to just have RegAccumLambda1/2 and RegAccumOp1/2 and have FMA be another case object?

**Review by mattfel1 — COMMENTED** [review]
Looks reasonable to me.  Should I get on this branch and start implementing the backend?

File: src/spatial/node/Accumulator.scala (hunk 1 above)
Comment (dkoeplin) [reply src/spatial/node/Accumulator.scala:51]: Yeah we could structure it that way. I had no other use cases requiring 2 inputs at the moment, so I just have it as FMA for now. If there are more (or we find more) we can generalize.

**Comment by dkoeplin** [comment]
Confirmed working for non-buffered accumulator registers. Need to check for buffered cases

**Comment by mattfel1** [comment]
Merged pull request because it successfully implements the specialized accumulators.  See issue #85  and #88 which were raised incidentally after this work and not yet fixed

**Pull request merged**